feat(android): Support android

dprevost-LMI opened the pull request:
In progress...

dprevost-LMI committed:
Draft work support Call Android notification style

dprevost-LMI committed:
Add CallStyle support in native

Needed to upgrade to `androidx.core:core` to 1.10 where they support callStyle in `NotificationCompat`

dprevost-LMI commented on .gitignore:
```diff
@@ -621,3 +621,5 @@ flutter_export_environment.sh
 #tests_react_native
 /tests_react_native/playgrounds
 /tests_react_native/packages
+
+.history/*
```
to remove

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/invertase/notifee?pullRequest=1134) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/invertase/notifee?pullRequest=1134) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/invertase/notifee?pullRequest=1134) it.</sub>

dprevost-LMI committed:
Review AndroidCallType + review validation

dprevost-LMI committed:
weird tentative to add intent in callStyle

dprevost-LMI committed:
non working upgrade to androidx.core 1.10

dprevost-LMI committed:
Need to update room too for duplicate class error

Without room update we had errors like below
```
Duplicate class kotlin.collections.jdk8.CollectionsJDK8Kt found in modules kotlin-stdlib-1.8.0
```

dprevost-LMI committed:
fix ts error

dprevost-LMI committed:
Use actions in the style to create intent for the different buttons

dprevost-LMI committed:
remove title & summary non existing on the call style

dprevost-LMI committed:
Add more validation to the callStyle

dprevost-LMI committed:
Code review

dprevost-LMI committed:
Add `Android Call Style` example

dprevost-LMI committed:
Better working example!

dprevost-LMI committed:
answer & decline trigger foreground events

dprevost-LMI committed:
fix error on triggering call style notif twice

dprevost-LMI committed:
Simplify config to only pass the callType without pressActions

dprevost-LMI committed:
Use default answerAction

mikehardy commented:
Just to note I see these commits flying by in my notification queue and I'm excited to see this when you think it's ready
No rush of course, it's your effort :-)
Totally fine with dependency updates as needed of course, core is actually on 1.15 now so bumping to 1.10 is arguably not even enough

mikehardy commented:
No worries about spam, it's a drop in the bucket and it's pleasing to see notificaitons about a feature PR vs yet another "I can't compile NNN because of ABC" issue comment 😅 

>  I reached a working point, but having them in the NotificationAndroidStyleModel.java. I want to move that into its own class at some point.
> However, I must pass the NotificationModel pretty deep, which seems bad. If you have any suggestions about that, let me know

I'm a fan of "tested and working" over perfect, so I lean towards something that you know works first. Second, I'm not sure what exactly would be better, passing it down the way it is now doesn't seem so bad

The only thing I can see that makes me 🤔 is the inconsistent use of constants vs "magic numbers and strings" - I love the enums at the TS level and it would be nice to see constants or enums to match at the java level (and in the TS warning, which uses the numbers again vs the call types)

In general though, looks like it is shaping up. At first I thought this was just duplication of our full screen intent style (which is targeted for this use case as well) but as soon as I looked at the APIs in use it looks like this is just me being unaware of the problem area and that the APIs have moved on (as they always seem to for notifications...) to have specific call notification type for modern android. Nice to support it

dprevost-LMI committed:
Add `AndroidStyle.CALL` in error validation

dprevost-LMI commented:
For your information, I just linked my project with my PR, and the first draft is working well in a happy path scenario!
I hope to wrap this up in the upcoming weeks!

From Notifee example project it looks like this:
<img width="348" alt="image" src="https://github.com/user-attachments/assets/edbbc6aa-9d51-42c6-b9e7-64c598501122">

mikehardy commented:
That's wonderful!

dprevost-LMI committed:
Use java enum for call & style type + finish validation unit test

dprevost-LMI committed:
Add style md doc

dprevost-LMI committed:
Code review

dprevost-LMI committed:
Better separation of concerns with `CallStyleNotificationPendingIntent`

dprevost-LMI committed:
Fix linting problems

dprevost-LMI committed:
add it for incoming call style + fix command type in contribution

dprevost-LMI committed:
fix const enum causing problem importing `AndroidCallType`

dprevost-LMI committed:
remove forgotten useless code + final code review

dprevost-LMI commented:
Do I need to create a feature issue for this PR, or the PR only is OK?

dprevost-LMI committed:
add missing coverage

dprevost-LMI committed:
Fixes on `yarn format:all:check`

dprevost-LMI committed:
Add `yarn format:all` to CONTRIBUTING.md

dprevost-LMI commented:
Hey there, I was wondering if I could help with anything regarding the next steps on that PR. Thanks!

mikehardy commented:
Sorry for the slow pace - progress here is happening it's just not visible (yet)

One of the biggest barriers to testing all-things-Notifee is that so much of it is in response to cloud messaging, and that necessarily requires a firebase project setup, an apple development account with messaging certs and keys setup and connected and everything, and that hadn't been done except for "borrowing" personal accounts for temporary / local unshareable one-off testing.

I'm nearly done with the process of having it set up in a permanent + general way, to unblock all testing, which then unblocks all this new feature + and foreground/background event-handling issue backlog

dprevost-LMI commented:
Great, thanks for the info. Your astonishing work and dedication are just amazing! Thanks again @mikehardy

dprevost-LMI committed:
Remove previous not needed changes anymore

dprevost-LMI commented:
Rebased! Ready again!

wilkinson4 commented:
Playing around with this locally, and I'm noticing that the heads up notification is visible for 5ish seconds, then disappears into the notification drawer. Am I missing something?

wilkinson4 commented:
> Playing around with this locally, and I'm noticing that the heads up notification is visible for 5ish seconds, then disappears into the notification drawer. Am I missing something?

After adding the `fullScreenAction` property with just an `id: default` key/value that changed the behavior of the notification

hevi08 commented:
@mikehardy could you please merge this PR ? Thanks

dprevost-LMI committed:
Working case with verification text and icon

dprevost-LMI commented:
Support for the verification text and icon was also added in 4798fdccc50cca19914777a89576392c376fcd74
<img width="319" alt="image" src="https://github.com/user-attachments/assets/a679f205-e3be-4b9f-9384-6e326e6d4fbf" />

dprevost-LMI committed:
Unit tests for new verification properties in call style

dprevost-LMI committed:
Fix docs

Irfanwani commented:
Any updates on this PR?

Irfanwani commented:
I tried using this PR (built the package locally from your fork). Call type notification is not having any effect, no buttons are shown. 
Are there any other changes required to make it work.
(As per the code, it doesn't see so, but in case i am missing something here)

Irfanwani commented:
btw, here is my implmentation:
```
 const channelId = await notifee.createChannel({
      id: 'call-channel',
      name: 'Call Notifications',
      importance: AndroidImportance.HIGH,
      visibility: AndroidVisibility.PUBLIC,
      sound: 'ringtone',
    });

    await notifee.displayNotification({
      // title: `${payload.name}`,
      // body: `${payload.number}`,
      id: 'call_notification',
      android: {
        channelId,
        smallIcon: 'ic_small_icon',
        category: AndroidCategory.CALL,
        importance: AndroidImportance.HIGH,
        visibility: AndroidVisibility.PUBLIC,
        loopSound: true,
        ongoing: true,
        autoCancel: false,
        sound: 'ringtone',
        style: {
          type: AndroidStyle.CALL,
          person: { name: "John Doe" },
          callTypeActions: {
            callType: AndroidCallType.INCOMING,
            answerAction: {
              pressAction: {
                id: 'accept',
                launchActivity: 'default',
              },
            },
            declineAction: {
              pressAction: {
                id: 'reject',
              },
            },
          },
          verification: { text: payload.number },
        },
        asForegroundService: true,
        foregroundServiceTypes: [
          AndroidForegroundServiceType.FOREGROUND_SERVICE_TYPE_CONNECTED_DEVICE,
        ],
        lightUpScreen: true,
        pressAction: {
          id: 'default',
        },
        fullScreenAction: {
          id: 'default',
          launchActivity: 'default',
        },
      },
    });

```

Irfanwani commented:
Seems the package is not having the updated code from PR.

Irfanwani commented:
@dprevost-LMI I am not able to use your fork. Can you please provide steps on how to use it in a react native app. I tried manually building but once used in the app, the app was failing to build. 
Also, this repo (your fork) cannot be directly installed via npm.

dprevost-LMI commented:
> @dprevost-LMI I am not able to use your fork. Can you please provide steps on how to use it in a react native app. I tried manually building but once used in the app, the app was failing to build. Also, this repo (your fork) cannot be directly installed via npm.

It was not meant to be used as a release package, but to be merged. Unfortunately, this one will need to be figured out without me 😉

Irfanwani commented:
FInally, packed it locally and used in my app. Working now. Thanks for this great work. 
I was about to start building this on my own (and that was definitely going to take a lot of time as i don't have much knowledge about native development).

Thanks for saving me a lot of time.
